Notebook created by [Nikolaos Tsopanidis](https://github.com/tSopermon)

# RAG System: Data Ingestion

Data ingestion is the first step in building a RAG system that will provide reliable data source to an LLM and enhance the capabilities of a pre-trained LLM.

For this task we need to import text-based documents referring to concert tours of 2025-2026 and store the information. The user will be able to ask questions to the LLM and receive information based on the reliable knowledge source which was created.

## Data ingestion process
In this process we will be able to load external documents and store them in a vector database for retrieval. The steps followed are:
 1. **Load**: Loading data into documents  
 2. **Split**: Splitting data into manageable chunks  
 3. **Embed**: Creating document embeddings  
 4. **Store**: Storing embeddings into a vector database 

<div style="text-align: center;">

![Alt](https://miro.medium.com/v2/resize:fit:720/format:webp/1*wHqtILSjqYsF6RnDq2CJDA.png)

Source: [medium.com - Amina Javaid](https://medium.com/@aminajavaid30/building-a-rag-system-the-data-ingestion-pipeline-d04235fd17ea)

</div>

## LangChain Implementation
**LangChain** is an open source framework used commonly in GenAI applications. Not only we can build apps, but we can use LangSmith, introduced by LangChain, to monitor LLMs, debug and evaluate code.

* First, we will install the required packages for the project

In [ ]:
%pip install langchain-community langchain-experimental langchain-huggingface faiss-cpu tiktoken sentence_transformers chromadb

  Using cached attrs-25.3.0-py3-none-any.whl.metadata (10 kB)
  Using cached httpx-0.28.1-py3-none-any.whl.metadata (7.1 kB)
  Using cached anyio-4.9.0-py3-none-any.whl.metadata (4.7 kB)
  Using cached h11-0.14.0-py3-none-any.whl.metadata (8.2 kB)
  Using cached jsonpointer-3.0.0-py2.py3-none-any.whl.metadata (2.3 kB)
  Using cached sniffio-1.3.1-py3-none-any.whl.metadata (3.9 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 16.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 30.7/30.7 MB 23.9 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 19.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 15.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 21.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.4/10.4 MB 23.4 MB/s eta 0:00:0000:010:01
Using cached attrs-25.3.0-py3-none-any.whl (63 kB)
Using cached httpx-0.28.1-py3-none-any.whl (73 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━

### **1. Data Loading**
Data extraction and convertion into a suitable format

#### **Loaders**
LangChain's loaders facilitate data ingestion and preprocessing from various sources. Data is loaded into a document object which includes the text content and the associated metadata. Loaders can also perform tokenization, normalization and format convertion, preparing data for LLMs.

In [5]:
"""
Loading text documents using LangChain Community.
"""
from langchain_community.document_loaders import TextLoader

file_path = "data/my_document.txt" # example file path
loader = TextLoader(file_path) # Load the document
documents = loader.load() # Load the documents
print(f"Loaded {len(documents)} document(s) from {file_path}")

Loaded 1 document(s) from data/my_document.txt


##### **Metadata** can prove useful when building an app where users need to retrieve information along with source or page location.

In [9]:
# Metadata created by the loader
for doc in documents:
    print(f"Document ID: {doc.metadata['source']}")
    print(f"Document Content: {doc.page_content[:100]}...")  # Print first 100 characters
    print(f"Document Metadata: {doc.metadata}")  # Print all metadata

Document ID: data/my_document.txt
Document Content: COVENANT

 

06.09.2025 Doors: 7:30 pm, Petra's Theater, M. Merkouri, Petroupoli 132 31, Athens




...
Document Metadata: {'source': 'data/my_document.txt'}


### **2. Data Splitting**
When preprocessing data for a RAG system, its important to consider the data's characteristics, system requirements, and limitations. Large documents may won't be processed by an LLM due to context window limits. To handle this, data must be split into smaller chunks while preserving context for effective retrieval. The reasons are:
 1. **Manageability**: Smaller parts will fit better within the LLM's context window, ensuring easier handling
 2. **Embedding model compatability**: Embedded models are limited on the amount of data they can process at a time, chunking the data offers better alignmnet with the model's capacity.
 3. **Efficient retrieval** When user sends a query, they only need specific information, often found in smaller chunks. Retrieving smaller chunks will decrease the resources needed instead of processing an entire corpus.

#### **Splitters**
Splitters divide large text documents into smaller chunks, ensuring they fit within model constraints and can be processed individually. Chunk size must be chosen carefully to preserve context. Overlapping chunks help maintain continuity, while minimal overlap works for distinct topics, ensuring high-quality responses.

We can adjust the chunk size depending on the limitations of the embedding model in use. Embedding models can be found on platforms like [Hugging Face](https://huggingface.co/models?other=embeddings), where each has specific input and output token limits
* Types of splitters: 
    - Recursive Splitter, 
    - HTML Splitter, 
    - Markdown Splitter, 
    - Code Splitter, 
    - Token Splitter, 
    - Character Splitter, 
    - Semantic Chunker

In [25]:
from langchain_text_splitters import CharacterTextSplitter
"""
CharacterTextSplitter splits text into chunks based on character count.
https://api.python.langchain.com/en/latest/character/langchain_text_splitters.character.CharacterTextSplitter.html
"""
text_splitter = CharacterTextSplitter(
    separator="\n\n",  # Split on double newlines
    chunk_size=400,  # Maximum size of each chunk
    chunk_overlap=50,  # Overlap between chunks
    length_function=len,  # Function to calculate the length of the text
)

chunks = []
for doc in documents:
    texts = text_splitter.split_text(doc.page_content)
    chunks.extend(texts)

In [26]:
# print the first 5 chunks
for i, chunk in enumerate(chunks[:5]):
    print(f"Chunk {i+1}: {chunk}")
    print(f"Chunk {i+1} Length: {len(chunk)}")
    print("-" * 50)

Chunk 1: COVENANT

 

06.09.2025 Doors: 7:30 pm, Petra's Theater, M. Merkouri, Petroupoli 132 31, Athens

INFO
 

The legendary Covenant at Petra's Theatre – Saturday, September 6, 2025

 
“Covenant, one of the leading names in EBM and dark electronic music, is coming to Athens for a unique performance at the Petra's Theatre on Saturday, September 6, 2025.
Chunk 1 Length: 349
--------------------------------------------------
Chunk 2: With a career spanning over three decades, Covenant has established itself as one of the most influential bands in electronic music. From the iconic "Like Tears in Rain," "Sequencer," "Northern Light," and "United States of Mind" to their more recent releases, the Swedish band continues to redefine their sound, always maintaining their signature intensity and atmosphere.
Chunk 2 Length: 373
--------------------------------------------------
Chunk 3: With tracks like "Dead Stars," "Call the Ships to Port," "Bullet," "Theremin," and many more, Covenant prom

### **3. Embeddings**
Textual data is converted into vector embeddings for efficient retrieval in a vector database. These dense, multi-dimensional vectors capture the semantic meaning of words, phrases, or sentences. LLMs generate embeddings, where semantically similar words have closer vector representations. Embeddings, often created using transformer-based models, can also detect anomalies by identifying outliers. Models are available as open-source on **Hugging Face** or paid options like **OpenAI**.

In [28]:
# initialization of the HuggingFaceEmbeddings model

from langchain_community.embeddings import HuggingFaceEmbeddings
"""
Source: https://huggingface.co/BAAI/bge-m3

BG3-M3 offers:
* Multi-Functionality: It can simultaneously perform the three common retrieval functionalities of embedding model: dense 
  retrieval, multi-vector retrieval, and sparse retrieval.
* Multi-Linguality: It can support more than 100 working languages.
* Multi-Granularity: It is able to process inputs of different granularities, spanning from short sentences to long 
  documents of up to 8192 tokens.

Suggestions for retrieval pipeline in RAG (future work):
* Hybrid retrieval leverages the strengths of various methods, offering higher accuracy and stronger generalization 
  capabilities. A classic example: using both embedding retrieval and the BM25 algorithm. Now, you can try to 
  use BGE-M3, which supports both embedding and sparse retrieval. This allows you to obtain token weights (similar to 
  the BM25) without any additional cost when generate dense embeddings. To use hybrid retrieval, you can refer to 
  Vespa (https://github.com/vespa-engine/pyvespa/blob/master/docs/sphinx/source/examples/mother-of-all-embedding-models-cloud.ipynb) 
  and Milvus.
* As cross-encoder models, re-ranker demonstrates higher accuracy than bi-encoder embedding model. Utilizing the 
  re-ranking model (e.g., bge-reranker, bge-reranker-v2) after retrieval can further filter the selected text.
"""

embeddings = HuggingFaceEmbeddings(
    model_name="BAAI/bge-m3",
    model_kwargs={
        "device": "cuda:0"  # Use GPU if available
    },
)

# Create embeddings of text chunks
for i, chunk in enumerate(chunks):
    print(f"Text chunk {i+1}")
    print(50*"-")
    print(chunk, "\n")
    query_result = embeddings.embed_query(chunk)
    print(f"Embeddings: {query_result}\n\n")

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/123 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/15.8k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/54.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/687 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/2.27G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/444 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.27G [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.1M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/964 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/191 [00:00<?, ?B/s]

Text chunk 1
--------------------------------------------------
COVENANT

 

06.09.2025 Doors: 7:30 pm, Petra's Theater, M. Merkouri, Petroupoli 132 31, Athens

INFO
 

The legendary Covenant at Petra's Theatre – Saturday, September 6, 2025

 
“Covenant, one of the leading names in EBM and dark electronic music, is coming to Athens for a unique performance at the Petra's Theatre on Saturday, September 6, 2025. 

Embeddings: [-0.00028646906139329076, 0.017611045390367508, -0.021807052195072174, 0.01063980907201767, -0.031538307666778564, -0.03949350491166115, -0.03802397847175598, -0.02537214569747448, 0.005703607574105263, 0.039865341037511826, 0.058910466730594635, 0.06073920801281929, -0.026946675032377243, -0.010229569859802723, -0.010005369782447815, -0.004005433525890112, 0.0710362121462822, -0.01597648300230503, 0.022679125890135765, -0.01205003634095192, -0.028786499053239822, 0.03700431436300278, -0.03054121881723404, -0.019439278170466423, 0.002887723268941045, -0.013673410750

We will choose work with all-mpnet-base-v2 model on this project because BG3-M3 model'size is 2.27 GB, while all-mpnet-base-v2 model's size is 438 MB.

In [29]:
embeddings = HuggingFaceEmbeddings(
    model_name = "sentence-transformers/all-mpnet-base-v2",
    model_kwargs = {
        "device": "cuda:0"  # Use GPU if available
    },
)

for i, chunk in enumerate(chunks):
    print("Text chunk ", i)
    print("--------------")
    print(chunk, "\n")
    query_result = embeddings.embed_query(chunk)
    print("Embeddings", "\n", query_result, "\n\n")

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.4k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/571 [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/438M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/363 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Text chunk  0
--------------
COVENANT

 

06.09.2025 Doors: 7:30 pm, Petra's Theater, M. Merkouri, Petroupoli 132 31, Athens

INFO
 

The legendary Covenant at Petra's Theatre – Saturday, September 6, 2025

 
“Covenant, one of the leading names in EBM and dark electronic music, is coming to Athens for a unique performance at the Petra's Theatre on Saturday, September 6, 2025. 

Embeddings 
 [-0.018523583188652992, 0.039547909051179886, -0.022635875269770622, 0.005549675785005093, -0.025357000529766083, -0.03754705563187599, 0.009629444219172001, 0.004501411225646734, 0.042958665639162064, 0.0386798270046711, -0.015448768623173237, -0.00583527609705925, 0.004367603920400143, 0.055153410881757736, 0.0019223694689571857, -0.00444506760686636, 0.004783324897289276, 0.04002365469932556, -0.02069455198943615, -0.02304098755121231, -0.06687997281551361, -0.003655659966170788, -0.04108191281557083, -0.02796647883951664, 0.007509034592658281, 0.005759683437645435, -0.049546562135219574, 0.00233

### **4. Data Storage - Vector Stores**
Vector stores efficiently store and query high-dimensional vector data, making them essential for AI applications like recommendation systems and RAG. Unlike traditional databases, which organize data in rows and columns, vector stores handle unstructured data like text, images, or speech more effectively. They are widely used in machine learning and retrieval tasks to find semantically similar information.

Vector store structure:
 * Vectors/Embeddings
 * Metadata
 * Original data
 * Unique ID

Vector stores use two main operations for querying:
- **Cosine Similarity**: Identifies top-k similar vectors by comparing the input vector (e.g., product description) with stored vectors. Metadata retrieves relevant details, commonly used in recommendation systems.
- **Dot Product**: Another method for comparing vectors effectively.

For this project, we will create a locally managed vector database suing [ChromaDB](https://github.com/chroma-core/chroma). Cloud-based services come with associated cloud sercice costs. Chroma is free and can run locally or on a remote server and it is ideal for flexible, LLM-focused tasks.

In [32]:
from langchain_community.vectorstores import Chroma
"""
https://api.python.langchain.com/en/latest/modules/indexes/vectorstores/langchain_community.vectorstores.chroma.Chroma.html
"""
vector_store = Chroma.from_texts(
    texts=chunks,
    embedding=embeddings,
    persist_directory="data/chroma_db",  # Directory to persist the database
    collection_name="my_collection"  # Name of the collection
)

## **References**
* https://medium.com/@aminajavaid30/building-a-rag-system-the-data-ingestion-pipeline-d04235fd17ea
* https://medium.com/@laddhaakshatrai/how-to-perform-data-ingestion-with-langchain-day-12-100-f11288d7ae99
* https://huggingface.co/sentence-transformers/all-mpnet-base-v2
* https://huggingface.co/mrhimanshu/finetuned-bge-m3